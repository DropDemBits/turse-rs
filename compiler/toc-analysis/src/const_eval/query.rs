//! Const eval query implementations

use std::{convert::TryInto, sync::Arc};

use toc_hir::{
    expr,
    symbol::{DefId, Mutability},
};

use crate::{
    const_eval::{
        errors::{ErrorKind, NotConst},
        ops::ConstOp,
        ConstError, ConstInt,
    },
    ty,
};

use super::{db::ConstEval, Const, ConstResult, ConstValue, EvalParams};

pub(crate) fn evaluate_const(
    db: &dyn ConstEval,
    expr: Const,
    params: EvalParams,
) -> ConstResult<ConstValue> {
    #[derive(Debug)]
    enum Eval {
        Expr(expr::ExprId),
        Op(ConstOp, toc_span::Span),
    }

    let (package_id, body_id) = match expr {
        Const::Value(v) => return Ok(v),
        Const::Error(err) => return Err(err),
        Const::Unevaluated(package, body) => (package, body),
        Const::UnevaluatedExpr(package, expr) => (package, expr.0),
    };

    let package = db.package(package_id);
    let body = package.body(body_id);
    let span_map = &package;

    let root_expr = match expr {
        Const::Unevaluated(_, _) => match &body.kind {
            toc_hir::body::BodyKind::Stmts(..) => {
                return Err(ConstError::new(
                    ErrorKind::NotConstExpr(NotConst::Expr),
                    body.span.lookup_in(span_map),
                ))
            }
            toc_hir::body::BodyKind::Exprs(expr) => *expr,
        },
        Const::UnevaluatedExpr(_, expr) => expr.1,
        _ => unreachable!(),
    };
    let in_module = db.inside_module((package_id, body_id, root_expr).into());

    // Do the actual evaluation, as a stack machine
    let mut eval_stack = vec![Eval::Expr(root_expr)];
    let mut operand_stack = vec![];

    loop {
        let local_expr = match eval_stack.pop() {
            Some(Eval::Expr(expr)) => expr,
            Some(Eval::Op(op, span)) => {
                // Perform operation
                let result = op
                    .evaluate(&mut operand_stack, params.allow_64bit_ops)
                    .map_err(|err| err.change_span(span))?;
                operand_stack.push(result);
                continue;
            }
            None => break,
        };

        let expr = body.expr(local_expr);
        let expr_span = expr.span.lookup_in(span_map);

        // Regarding enum field accesses:
        // We have access to the typing context now, so it should be easy to support enums & the like
        match &expr.kind {
            expr::ExprKind::Missing => {
                // Bail out
                return Err(ConstError::new(ErrorKind::MissingExpr, expr_span));
            }
            expr::ExprKind::Literal(expr) => {
                // ???: How to deal with 32-bit vs 64-bit integers?
                // - Could yoink info from somewhere?
                // - Only need to know if ops need to be done in either 32 or 64 bit mode

                // Convert into a ConstValue
                let operand = match expr {
                    expr::Literal::Integer(v) => {
                        let v = ConstInt::from_unsigned(*v, params.allow_64bit_ops)
                            .map_err(|err| err.change_span(expr_span))?;
                        ConstValue::Integer(v)
                    }
                    expr::Literal::Real(v) => ConstValue::Real(*v),
                    expr::Literal::Boolean(v) => ConstValue::Bool(*v),
                    expr::Literal::Char(v) => ConstValue::Char(*v),
                    expr::Literal::CharSeq(v) => ConstValue::CharN(Arc::new(v.clone())),
                    expr::Literal::String(v) => ConstValue::String(Arc::new(v.clone())),
                };

                operand_stack.push(operand);
            }
            expr::ExprKind::Binary(expr) => {
                // Push both expression operands and the operation
                let (op, span) = (expr.op.try_into()?, expr.op.span());
                eval_stack.push(Eval::Op(op, span.lookup_in(span_map)));
                eval_stack.push(Eval::Expr(expr.rhs));
                eval_stack.push(Eval::Expr(expr.lhs));
            }
            expr::ExprKind::Unary(expr) => {
                // Push expr operand & operator
                let (op, span) = (expr.op.try_into()?, expr.op.span());
                eval_stack.push(Eval::Op(op, span.lookup_in(span_map)));
                eval_stack.push(Eval::Expr(expr.rhs));
            }
            expr::ExprKind::Name(name) => {
                // May or may not reference a constant expression
                match name {
                    expr::Name::Name(binding) => {
                        // Resolve to canonical def first so that we aren't looking at any exports
                        let canonical_def = match package.binding_resolve(*binding) {
                            toc_hir::symbol::Resolve::Def(local_def) => {
                                db.resolve_def(DefId(package_id, local_def)).ok()
                            }
                            toc_hir::symbol::Resolve::Err => None,
                        }
                        .ok_or_else(|| {
                            // Not a const expr
                            ConstError::new(
                                ErrorKind::NotConstExpr(NotConst::Binding(package_id, *binding)),
                                expr_span,
                            )
                        })?;

                        let package_id = canonical_def.package();
                        let package = db.package(package_id);

                        let body = db.item_of(canonical_def).and_then(|item| {
                            match &package.item(item.1).kind {
                                toc_hir::item::ItemKind::ConstVar(cv)
                                    if matches!(cv.mutability, Mutability::Const) =>
                                {
                                    cv.init_expr
                                }
                                _ => None,
                            }
                        });

                        let body = match body {
                            Some(body) => body,
                            None => {
                                // Not a const expr
                                return Err(ConstError::new(
                                    ErrorKind::NotConstExpr(NotConst::Def(canonical_def)),
                                    expr_span,
                                ));
                            }
                        };

                        let value =
                            db.evaluate_const(Const::from_body(package_id, body), params)?;

                        // Check that the produced value matches the real type of the def
                        let left_ty = db.type_of(canonical_def.into());
                        let right_ty = db.type_of((package_id, body).into());

                        // FIXME: add tests with opaque tys once module exports are in const eval
                        let left_ty = left_ty
                            .peel_opaque(db.up(), in_module)
                            .peel_aliases(db.up());
                        let right_ty = right_ty
                            .peel_opaque(db.up(), in_module)
                            .peel_aliases(db.up());

                        if !ty::rules::is_assignable(db, left_ty, right_ty) {
                            // Wrong types
                            let span = package.body(body).span.lookup_in(&package);
                            return Err(ConstError::new(ErrorKind::WrongResultType, span));
                        }

                        // If bounds of `left` is known, check if `right` is in the range
                        if let Some((min, max)) =
                            Option::zip(left_ty.min_int_of(db).ok(), left_ty.max_int_of(db).ok())
                        {
                            // Since right is assignable into left, we can treat right as ConstInt
                            let as_ordinal = value.ordinal().ok_or_else(|| {
                                // Definitely the wrong type
                                let span = package.body(body).span.lookup_in(&package);
                                ConstError::new(ErrorKind::WrongResultType, span)
                            })?;

                            if !(min..=max).contains(&as_ordinal) {
                                // Is outside of value range
                                let span = package.body(body).span.lookup_in(&package);
                                return Err(ConstError::new(ErrorKind::OutsideRange, span));
                            }
                        }

                        // Cast into the canonical value
                        let value = match left_ty.to_base_type(db.up()).kind(db.up()) {
                            ty::TypeKind::Char => value.cast_into_char().map_or_else(
                                |err| {
                                    // Definitely the wrong type
                                    let span = package.body(body).span.lookup_in(&package);
                                    Err(err.change_span(span))
                                },
                                |v| Ok(ConstValue::Char(v)),
                            )?,
                            _ => value,
                        };

                        // Push (cached) value to the operand stack
                        operand_stack.push(value);
                    }
                    expr::Name::Self_ => {
                        // Never a const expr
                        // TODO: Use the self's associated def_id
                        return Err(ConstError::new(
                            ErrorKind::NotConstExpr(NotConst::Expr),
                            expr_span,
                        ));
                    }
                }
            }
            expr::ExprKind::Field(expr) => {
                // Possible things:
                // - module export (constvar reference)
                // - enum variants (enum variant value)
                let lhs_expr = (package_id, body_id, expr.lhs);
                let lhs_tyref = db
                    .type_of(lhs_expr.into())
                    .peel_opaque(db.up(), in_module)
                    .peel_aliases(db.up());

                match lhs_tyref.kind(db.up()) {
                    ty::TypeKind::Enum(_, variants) => {
                        // Enum variants
                        let Some(package_id) = variants.first().map(|def| def.package()) else {
                            return Err(ConstError::new(
                                ErrorKind::NoFields(*expr.field.item()),
                                expr.field.span().lookup_in(&package),
                            ));
                        };
                        let package = db.package(package_id);

                        // Get variant ordinal
                        let ordinal = variants
                            .iter()
                            .enumerate()
                            .find_map(|(idx, def_id)| {
                                (package.local_def(def_id.1).name == *expr.field.item())
                                    .then_some(idx)
                            })
                            .ok_or_else(|| {
                                ConstError::new(
                                    ErrorKind::NoFields(*expr.field.item()),
                                    expr.field.span().lookup_in(&package),
                                )
                            })?;

                        operand_stack.push(ConstValue::EnumVariant(lhs_tyref, ordinal));
                    }
                    _ => {
                        // Defer to normal field lookup
                        // FIXME: Handle const-eval field lookups for modules
                        // Note: This will expose us to evaluation cycles, so those need to be handled as well
                        // when the time comes
                        return Err(ConstError::new(
                            ErrorKind::NotConstExpr(NotConst::Expr),
                            expr_span,
                        ));
                    }
                }
            }
            expr::ExprKind::Call(_) => {
                // There are some functions which are allowed to be const-fns, but they're all builtins
                // It's okay to treat it as not const-evaluable
                return Err(ConstError::new(
                    ErrorKind::NotConstExpr(NotConst::Expr),
                    expr_span,
                ));
            }
            _ => {
                // The rest of the exprs aren't ever evaluable at compile time
                return Err(ConstError::new(
                    ErrorKind::NotConstExpr(NotConst::Expr),
                    expr_span,
                ));
            }
        }
    }

    // Should be one value remaining
    let result = operand_stack
        .pop()
        .expect("All values popped off of operand stack");
    assert!(operand_stack.is_empty());

    Ok(result)
}
