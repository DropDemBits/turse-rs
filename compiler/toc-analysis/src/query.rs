//! Analysis query implementation

use toc_reporting::{CompileResult, MessageBundle};

#[salsa::tracked(jar = crate::db::AnalysisJar)]
pub(crate) fn analyze_libraries(db: &dyn crate::db::HirAnalysis) -> CompileResult<()> {
    let mut messages = MessageBundle::default();

    // FIXME: Report cyclic dep errors
    let source_graph = toc_source_graph::source_graph(db.upcast_to_source_graph_db())
        .as_ref()
        .unwrap();
    let res = toc_hir_lowering::lower_source_graph(db.upcast_to_lowering_db());
    res.bundle_messages(&mut messages);

    for &library in source_graph.all_libraries(db.upcast_to_source_graph_db()) {
        db.typecheck_library(library).bundle_messages(&mut messages);
        db.lint_library(library).bundle_messages(&mut messages);
    }

    CompileResult::new((), messages)
}
