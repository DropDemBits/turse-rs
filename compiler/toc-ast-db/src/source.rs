//! Source file interpretation queries

use std::{
    collections::{BTreeSet, VecDeque},
    sync::Arc,
};

use toc_parser::ExternalLinks;
use toc_reporting::CompileResult;
use toc_span::{FileId, Span};
use toc_vfs_db::SourceFile;

use crate::Db;

/// What other files a given file refers to
#[salsa::tracked]
pub fn file_links(db: &dyn Db, source: SourceFile) -> Arc<toc_parser::ExternalLinks> {
    let mut links = ExternalLinks::default();
    let deps = crate::parse_depends(db, source);
    for dep in deps.result().dependencies() {
        // Get the target file
        let child =
            toc_vfs_db::resolve_path(db.up(), source.path(db.up()), dep.relative_path.clone());

        links.bind(dep.link_from.clone(), child.into());
    }

    Arc::new(links)
}

/// Parses the given file
#[salsa::tracked]
pub fn parse_file(db: &dyn Db, source: SourceFile) -> CompileResult<toc_parser::ParseTree> {
    // FIXME: If a load error is present, then add it to the parse result / create a new one
    let file_id: FileId = source.path(db.up()).into();
    toc_parser::parse(source.contents(db.up())).remap_spans(|range| Span::new(file_id, range.0))
}

/// Validates the file according to grammar validation rules
#[salsa::tracked]
pub fn validate_file(db: &dyn Db, source: SourceFile) -> CompileResult<()> {
    let file_id: FileId = source.path(db.up()).into();
    let cst = crate::parse_file(db, source);
    toc_validate::validate_ast(cst.result().syntax())
        .remap_spans(|range| Span::new(file_id, range.0))
}

/// Parse out the dependencies of a file
#[salsa::tracked]
pub fn parse_depends(
    db: &dyn Db,
    source: SourceFile,
) -> CompileResult<Arc<toc_parser::FileDepends>> {
    let file_id: FileId = source.path(db.up()).into();
    let cst = crate::parse_file(db, source);
    toc_parser::parse_depends(cst.result().syntax())
        .remap_spans(|range| Span::new(file_id, range.0))
}

/// Gets the [`ExternalLink`](toc_parser::ExternalLink)'s corresponding file
#[salsa::tracked]
pub fn file_link_of(
    db: &dyn Db,
    source: SourceFile,
    link: toc_parser::ExternalLink,
) -> Option<FileId> {
    crate::file_links(db, source).links_to(link)
}

/// Gets the set of all the transient file dependencies of `root` (excluding itself)
#[salsa::tracked]
pub fn reachable_files(db: &dyn Db, root: SourceFile) -> Arc<BTreeSet<SourceFile>> {
    let mut files = BTreeSet::default();
    let mut pending_queue = VecDeque::default();

    pending_queue.push_back(root);

    while let Some(current_file) = pending_queue.pop_front() {
        // Skip if we've already explored this file
        if files.contains(&current_file) {
            continue;
        }
        files.insert(current_file);

        // add all of this file's linked bits to the queue
        pending_queue.extend(
            crate::file_links(db, current_file)
                .all_links()
                .map(|path| toc_vfs_db::source_of(db.up(), path.into_raw())),
        );
    }

    files.remove(&root);

    Arc::new(files)
}

/// Gets the set of all the transient file imports of `root` (excluding itself)
#[salsa::tracked]
pub fn reachable_imported_files(db: &dyn Db, root: SourceFile) -> Arc<BTreeSet<SourceFile>> {
    let mut files = BTreeSet::default();
    let mut pending_queue = VecDeque::default();

    pending_queue.push_back(root);

    while let Some(current_file) = pending_queue.pop_front() {
        // Skip if we've already explored this file
        if files.contains(&current_file) {
            continue;
        }
        files.insert(current_file);

        // add all of the file's import linked bits to the queue
        pending_queue.extend(
            crate::file_links(db, current_file)
                .import_links()
                .map(|path| toc_vfs_db::source_of(db.up(), path.into_raw())),
        );
    }

    files.remove(&root);

    Arc::new(files)
}
