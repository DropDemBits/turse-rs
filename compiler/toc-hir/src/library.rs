//! Library structure definitions

use std::sync::Arc;

use indexmap::IndexMap;
use la_arena::{Arena, Idx};
use toc_span::FileId;

use crate::{
    body, item,
    span::{HasSpanTable, SpanTable, Spanned},
    symbol::{self, ResolutionMap},
    ty,
};

pub use crate::library_graph::LibraryId;

/// A reference to a library node in a specific library
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct InLibrary<T>(pub LibraryId, pub T);

impl<T> InLibrary<T> {
    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> InLibrary<U> {
        InLibrary(self.0, f(self.1))
    }

    pub fn library(self) -> LibraryId {
        self.0
    }

    pub fn item(self) -> T {
        self.1
    }
}

pub trait WrapInLibrary: Copy {
    type Output: Copy;

    /// Wraps self in the context of a library
    fn in_library(self, library: LibraryId) -> InLibrary<Self::Output>;
}

impl<T> WrapInLibrary for T
where
    T: Copy,
{
    type Output = T;

    fn in_library(self, library: LibraryId) -> InLibrary<Self::Output> {
        InLibrary(library, self)
    }
}

/// A `Library` represents a logical collection of files.
///
/// It is a conceptual group of files / units that are accessible from a
/// specific root file. For example, all of the standard library files are
/// grouped together under one `Library`. Similarly, the file provided during
/// compilation serves as a root for a `Library`.
///
/// `Library` stores the arena for all library local entities
/// (`Item`, `Body`, `DefInfo`, and `Type`).
/// `Library` also stores a mapping between the library files, and the
/// associated root `ItemId`s.
#[derive(PartialEq, Eq)]
pub struct Library {
    /// Map between library files and root items
    pub root_items: IndexMap<FileId, item::ItemId>,
    /// Table of all interned spans
    pub(crate) span_map: SpanTable,
    /// Table of all interned types
    pub(crate) type_map: ty::TypeTable,
    pub(crate) items: Arena<item::Item>,
    pub(crate) defs: symbol::DefInfoTable,
    pub(crate) bodies: Arena<body::Body>,
    pub(crate) resolve_map: ResolutionMap,
}

impl std::fmt::Debug for Library {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Library")
            .field("root_items", &self.root_items)
            .finish_non_exhaustive()
    }
}

impl Library {
    pub fn body(&self, body_id: body::BodyId) -> &body::Body {
        &self.bodies[Idx::from(body_id)]
    }

    pub fn item(&self, item_id: item::ItemId) -> &item::Item {
        &self.items[Idx::from(item_id)]
    }

    pub fn module_item(&self, module_id: item::ModuleId) -> item::ModuleLike<'_> {
        match &self.item(module_id.item_id()).kind {
            item::ItemKind::Module(module) => item::ModuleLike::Module(module),
            _ => unreachable!("not a module-like"),
        }
    }

    pub fn local_def(&self, local_def: symbol::LocalDefId) -> &symbol::DefInfo {
        self.defs.get_info(local_def)
    }

    pub fn lookup_type(&self, type_id: ty::TypeId) -> &ty::Type {
        self.type_map.lookup_type(type_id)
    }

    pub fn binding_resolve(&self, binding: Spanned<symbol::Symbol>) -> symbol::Resolve {
        self.resolve_map
            .binding_resolves
            .get(&binding)
            .copied()
            .unwrap_or(symbol::Resolve::Err)
    }

    pub fn def_resolve(&self, local_def: symbol::LocalDefId) -> symbol::DefResolve {
        self.resolve_map
            .def_resolves
            .get(local_def)
            .copied()
            .unwrap_or(symbol::DefResolve::Canonical)
    }

    pub fn local_defs(&self) -> impl Iterator<Item = symbol::LocalDefId> + '_ {
        self.defs.iter().map(|(id, _)| id)
    }

    pub fn body_ids(&self) -> Vec<body::BodyId> {
        self.bodies
            .iter()
            .map(|(idx, _)| body::BodyId(idx))
            .collect()
    }
}

impl HasSpanTable for Library {
    fn span_table(&self) -> &SpanTable {
        &self.span_map
    }
}

/// Lowered [`Library`].
///
/// Data is wrapped inside of an [`Arc`], so it is trivially cloneable.
///
/// [`Library`]: crate::library::Library
pub type LoweredLibrary = Arc<Library>;
