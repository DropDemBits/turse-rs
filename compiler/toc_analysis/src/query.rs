//! Analysis query implementation

use toc_reporting::CompileResult;

pub(crate) fn analyze_libraries(db: &dyn crate::db::HirAnalysis) -> CompileResult<()> {
    let mut messages = vec![];

    let res = db.library_graph();
    let lib_graph = res.result();
    res.bundle_messages(&mut messages);

    for (_, library) in lib_graph.library_roots() {
        db.typecheck_library(library).bundle_messages(&mut messages);
    }

    CompileResult::new((), messages)
}
