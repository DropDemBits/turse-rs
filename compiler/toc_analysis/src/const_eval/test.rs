use std::cell::RefCell;

use toc_ast_db::source::SourceParser;
use toc_hir::library::{LibraryId, LoweredLibrary};
use toc_hir_db::HirDatabase;
use toc_reporting::{MessageSink, ReportMessage};
use toc_salsa::salsa;
use toc_vfs::query::VfsDatabaseExt;
use unindent::unindent;

use crate::{const_eval::Const, db::ConstEval};

#[track_caller]
fn assert_const_eval(source: &str) {
    insta::assert_snapshot!(insta::internals::AutoName, do_const_eval(source), source);
}

/// Expr version
#[track_caller]
fn assert_const_eval_expr(expr: &str) {
    let source = format!("const _ := {}", expr);
    insta::assert_snapshot!(insta::internals::AutoName, do_const_eval(&source), expr);
}

macro_rules! for_all_const_exprs {
    ($($src:literal)+) => {
        $(assert_const_eval_expr($src);)+
    };
}

fn do_const_eval(source: &str) -> String {
    let mut db = TestDb::default();
    let root_file = db.vfs.intern_path("src/main.t".into());
    db.update_file(root_file, Some(source.into()));

    let source_roots = toc_ast_db::source::SourceRoots::new(vec![root_file]);
    db.set_source_roots(source_roots);

    let library_id = db.library_graph().result().library_of(root_file);
    let library = db.library(library_id);

    // Eagerly evaluate all of the available const vars
    let const_evaluator = ConstEvaluator {
        db: &db,
        library_id,
        library: library.clone(),
        reporter: Default::default(),
        results: Default::default(),
    };

    toc_hir::visitor::preorder_visit_library(&library, &const_evaluator);

    struct ConstEvaluator<'db> {
        db: &'db TestDb,
        library_id: LibraryId,
        library: LoweredLibrary,
        reporter: RefCell<MessageSink>,
        results: RefCell<String>,
    }

    impl toc_hir::visitor::HirVisitor for ConstEvaluator<'_> {
        fn visit_constvar(&self, id: toc_hir::item::ItemId, item: &toc_hir::item::ConstVar) {
            let def_id = self.library.item(id).def_id;
            let body = if let Some(body) = item.tail.init_expr() {
                body
            } else {
                return;
            };

            let eval_res = self
                .db
                .evaluate_const(Const::from_body(self.library_id, body), Default::default());

            let name = {
                let def_info = self.library.local_def(def_id);
                format!(
                    "{:?}@{:?}",
                    def_info.name.item(),
                    def_info.name.span().lookup_in(&self.library.span_map)
                )
            };

            let results = match eval_res {
                Ok(v) => format!("{} -> {:?}\n", name, v),
                Err(err) => {
                    let text = format!("{} -> {:?}\n", name, err);
                    err.report_to(self.db, &mut *self.reporter.borrow_mut());
                    text
                }
            };

            self.results.borrow_mut().push_str(&results);
        }
    }

    let ConstEvaluator {
        reporter, results, ..
    } = const_evaluator;
    let reporter = reporter.into_inner();
    let results = results.into_inner();

    // Errors are bundled into the const error context
    stringify_const_eval_results(&results, &reporter.finish())
}

fn stringify_const_eval_results(results: &str, messages: &[ReportMessage]) -> String {
    // Pretty print const eval ctx
    // Want:
    // - Evaluation state of all accessible DefIds
    let mut s = format!("{}\n", results);

    // Pretty print the messages
    for err in messages.iter() {
        s.push_str(&format!("{}\n", err));
    }

    s
}

#[salsa::database(
    toc_vfs::query::FileSystemStorage,
    toc_ast_db::source::SourceParserStorage,
    toc_hir_db::HirDatabaseStorage,
    crate::db::TypeInternStorage,
    crate::db::TypeDatabaseStorage,
    crate::db::ConstEvalStorage,
    crate::HirAnalysisStorage
)]
#[derive(Default)]
struct TestDb {
    storage: salsa::Storage<Self>,
    vfs: toc_vfs::Vfs,
}

impl salsa::Database for TestDb {}

impl toc_vfs::HasVfs for TestDb {
    fn get_vfs(&self) -> &toc_vfs::Vfs {
        &self.vfs
    }
}

#[test]
fn complex_arithmetic_expr() {
    assert_const_eval_expr("1 + 2 * 3 div 3 - 4 + 5");
}

#[test]
fn arithmetic_const_ops() {
    // All operations should be evaluated
    for_all_const_exprs![
        "1 + 1"
        "10 - 2"
        "2 * 5"
        "3 div 2"
        "-1"
        "+-1"

        "1 + 1.0"
        "10 - 2.0"
        "2 * 5.0"
        "3 div 2.0"
        "-1.0"
        "+-1.0"

        // RealDiv
        "1.0 / 2.0"
        "1 / 2"
        "1.0 / -2"

        // Mod
        " 3 mod 10"
        "-3 mod 10"
        " 3 mod -10"
        "-3 mod -10"

        " 3.0 mod 10.0"
        "-3.0 mod 10.0"
        " 3.0 mod -10.0"
        "-3.0 mod -10.0"

        " 3.5 mod 10.0   % 3.5"
        "-3.5 mod 10.0   % 6.5"
        " 3.5 mod -10.0   % -6.5"
        "-3.5 mod -10.0   % -3.5"

        " 3.0 mod 10.5   % 3"
        "-3.0 mod 10.5   % 7.5"
        " 3.0 mod -10.5   % -7.5"
        "-3.0 mod -10.5   % -3"

        // Rem
        " 3 mod 10"
        "-3 mod 10"
        " 3 mod -10"
        "-3 mod -10"

        " 3.0 mod 10.0"
        "-3.0 mod 10.0"
        " 3.0 mod -10.0"
        "-3.0 mod -10.0"

        // Exp
        "2 ** 4"
        "2 ** 0"
        "-2 ** 4"
        "-2 ** 0"

        "2.0 ** 4"
        "2.0 ** 0"
        "-2.0 ** -4"
        "-2.0 ** -2"
    ];
}

#[test]
fn logic_const_ops() {
    for_all_const_exprs![
        "false and false" // 0
        "false and true"  // 0
        "true and false"  // 0
        "true and true"   // 1

        "false or false"  // 0
        "false or true"   // 1
        "true or false"   // 1
        "true or true"    // 1

        "false xor false" // 0
        "false xor true"  // 1
        "true xor false"  // 1
        "true xor true"   // 0

        "false => false"  // 1
        "false => true"   // 1
        "true => false"   // 0
        "true => true"    // 1

        "not true"
        "not false"
    ];
}

#[test]
fn bitwise_const_ops() {
    for_all_const_exprs![
        "0 and 0"
        "0 and 1"
        "1 and 0"
        "1 and 1"

        "0 or 0"
        "0 or 1"
        "1 or 0"
        "1 or 1"

        "0 xor 0"
        "0 xor 1"
        "1 xor 0"
        "1 xor 1"

        "not 0"
        "not 1"

        // Applicable to negative numbers too
        "0 and -0"
        "0 and -1"
        "1 and -0"
        "1 and -1"

        "0 or -0"
        "0 or -1"
        "1 or -0"
        "1 or -1"

        "0 xor -0"
        "0 xor -1"
        "1 xor -0"
        "1 xor -1"

        "not -0"
        "not -1"

        "1 shl 3"
        "-1 shl 0"

        "24 shr 3"
        "-24 shr 3"

        // Wrapping around the mask
        "1 shl 32"
        "8 shr 32"
    ];
}

#[test]
fn real_promotion() {
    for_all_const_exprs![
        "1.0 + 1"
        "  1 + 1.0"
        "1.0 + 1.0"

        "1.0 - 1"
        "  1 - 1.0"
        "1.0 - 1.0"

        "1.0 * 1"
        "  1 * 1.0"
        "1.0 * 1.0"

        "1.0 div 1"
        "  1 div 1.0"
        "1.0 div 1.0"
    ];
}

#[test]
fn const_local_var_lookup() {
    assert_const_eval_expr(&unindent(
        r#"
    const a := 1
    const b := a + 1
    const c := b
    "#,
    ));
}

#[test]
fn error_div_by_zero() {
    // integer div
    for_all_const_exprs![
        "1 div 0"
        "1 div 0.0"
        "1.0 div 0"
        "1.0 div 0.0"
    ];

    // real div
    for_all_const_exprs![
        "1 / 0"
        "1 / 0.0"
        "1.0 / 0"
        "1.0 / 0.0"
    ];

    // modulus
    for_all_const_exprs![
        "1 mod 0"
        "1 mod 0.0"
        "1.0 mod 0"
        "1.0 mod 0.0"
    ];

    // remainder
    for_all_const_exprs![
        "1 rem 0"
        "1 rem 0.0"
        "1.0 rem 0"
        "1.0 rem 0.0"
    ];
}

#[test]
fn error_int_overflow_32bit() {
    // TODO: Test operations on 64-bit values

    for_all_const_exprs![
        // 64-bit literals should overflow right away
        "16#100000000"
        // Unsigned 32-bit literals shouldn't
        "16#80000000 % no overflow"
        "16#FFFFFFFF % no overflow"
        // Over add
        "16#ffffffff + 1"
        "-16#80000000 + -1"
        // Over sub
        "16#ffffffff - (-1)"
        "-16#80000000 - 1"
        "16#ffffffff - 16#ffffffff % no overflow"
        // Over mul (positive result)
        "16#80000000 * 2"
        "-16#80000000 * -1 % no overflow"
        // Over mul (negative result)
        "16#80000001 * -1"
        "-16#80000000 * 2"
        // Over idiv (positive result)
        "1e100 div 1"
        "1e100 div 1.0"
        // Over idiv (negative result)
        "16#FFFFFFFF div -1"
        "16#80000001 div -1"
        // Over negate
        "-16#FFFFFFFF"
        "--16#80000000 % no overflow"
        // Over identity (no overflow)
        "+16#FFFFFFFF % no overflow"
        "+-16#80000000 % no overflow"

        // modulo is safe except for large inputs (overflow before then)
        // rem is safe except for large inputs (overflow before then)

        // Over Exp
        "2 ** 33"
        "2 ** 32"
        "(-2) ** 33"

        // Over shl
        "2 shl 31"
        // Any shl over negative integers will overflow
        "-1 shl 1"
    ];
}

#[test]
fn error_real_overflow() {
    for_all_const_exprs![
        // Over add
        "1e308 + 1e308"
        "-1e308 + (-1e308)"
        // Over sub
        "1e308 - (-1e308)"
        "-1e308 - 1e308"
        // Over mul
        "1e308 * 10"
        "-1e308 * 10"
        // Over idiv (checked in `error_int_overflow_{32,64}_bit`)
        // Over rdiv
        "1e308 / 0.1"
        "-1e308 / 0.1"
        // modulo is safe except for large inputs (overflow before then)
        // rem is safe except for large inputs (overflow before then)
        // Over exp
        "10.0 ** 309"
        "(-10.0) ** 309"
    ];
}

#[test]
fn error_arithmetic_wrong_types() {
    for_all_const_exprs![
        "1 + true"
        "1 - true"
        "1 * true"
        "1 div true"
        "false + 1"
        "false - 1"
        "false * 1"
        "false div 1"

        "1.0 + true"
        "1.0 - true"
        "1.0 * true"
        "1.0 div true"
        "false + 1.0"
        "false - 1.0"
        "false * 1.0"
        "false div 1.0"

        "-false"
        "+true"

        "1 / true"
        "1 mod true"
        "1 rem true"
        "1 ** true"
        "1.0 / true"
        "1.0 mod true"
        "1.0 rem true"
        "1.0 ** true"
    ];
}

#[test]
fn error_logical_wrong_types() {
    for_all_const_exprs![
        "1.0 and false"
        "1.0 or false"
        "1.0 xor false"
        "1.0 => false"

        "false and 1.0"
        "false or 1.0"
        "false xor 1.0"
        "false => 1.0"

        "not 1.0"

        "1 shl 1.0"
        "false shr 1"
    ];
}

#[test]
fn error_no_const_expr() {
    // Referencing a runtime-evaluated var
    assert_const_eval(&unindent(
        r#"
    var a := 1
    const b := a
    "#,
    ));

    // Referencing a runtime-evaluated const
    assert_const_eval(&unindent(
        r#"
    var a := 1
    const b := a
    const c := b
    "#,
    ));

    // Referencing `self`
    // TODO: Uncomment when `self` is lowered again
    /*
    assert_const_eval(&unindent(
        r#"
    const a := self
    "#,
    ));
    */
}

#[test]
fn error_propagation() {
    // Propagation of errors
    assert_const_eval(&unindent(
        r#"
    const a := 1 div 0
    const b := a
    "#,
    ));
}

#[test]
fn error_negative_int_exp() {
    assert_const_eval_expr("2 ** -1");
}

#[test]
fn error_negative_int_shift() {
    assert_const_eval_expr("2 shl -1");
    assert_const_eval_expr("2 shr -1");
}

#[test]
fn restrict_assign_type() {
    // Assignment type restriction is only checked on const use (for eval only)
    // Boolean is assignable into boolean
    assert_const_eval(r#"const a : boolean := false const _:=a"#);
    // Integer is not assignable into boolean
    assert_const_eval(r#"const a : boolean := 1 const _:=a"#);
    // Real is not assignable into boolean
    assert_const_eval(r#"const a : boolean := 1.0 const _:=a"#);

    // Boolean is not assignable into integers
    assert_const_eval(r#"const a : int := false const _:=a"#);
    // Integer is assignable into integer
    assert_const_eval(r#"const a : int := 1 const _:=a"#);
    // Real is not assignable into integers
    assert_const_eval(r#"const a : int := 1.0 const _:=a"#);

    // Boolean is not assignable into real
    assert_const_eval(r#"const a : real := false const _:=a"#);
    // Integers are assignable into reals (promoted)
    assert_const_eval(r#"const a : real := 1 const _:=a"#);
    // Real is assignable into real
    assert_const_eval(r#"const a : real := 1.0 const _:=a"#);
}

#[test]
fn unsupported_values() {
    assert_const_eval(r#"const a := "alphabet""#);
    assert_const_eval(r#"const a := 'fun times'"#);
    assert_const_eval(r#"const a := 'e'"#);
}

#[test]
fn unsupported_ops() {
    for_all_const_exprs![
        "1 > 1"
        "1 >= 1"
        "1 < 1"
        "1 <= 1"
        "1 = 1"
        "1 ~= 1"
        "1 in 1"
        "1 ~in 1"
    ];
}
