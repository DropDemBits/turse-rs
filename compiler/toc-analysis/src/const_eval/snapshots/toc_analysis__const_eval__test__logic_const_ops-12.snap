---
source: compiler/toc_analysis/src/const_eval/test.rs
expression: true xor true

---
"_"@(FileId(1), 6..7) -> Bool(false)


